rosserial_windows - package for building rosserial clients on Windows

kshehata opened the pull request:
I've added the rosserial_windows package for building rosserial_clients for Windows using Visual Studios. There's a tutorial on how to use the library on the ROS wiki http://wiki.ros.org/rosserial_windows/Tutorials/Hello%20World#preview

This is based on Hydro because my current environment is Hydro, but should be an easy port if not identical on Indigo.

Commit:
Adding rosserial_windows for Windows clients on rosserial.
Using Arduino as a starting point.

Commit:
Started making package for Windows.
Updated package definition and CMakelists
Brought in customizations for WindowsSocket from existing project.
Renamed directories.
Removed Arduino-specific elements.

Commit:
Updated make_libraries.py for Windows. Only really changed names.
Also removed ArduinoHardware.h which got missed in the last commit.

Commit:
Rewrote the Windows Socket interface for rosserial_windows.
The old code was a hack put in place. This is slightly better.
Also updated the license boilerplate.

Commit:
Updated the changelog to show just the 0.0.0 release for
rosserial_windows

Commit:
Removing leftovers from Arduino

Commit:
Cleaned up CMakeLists after removing the Arduino examples

Commit:
Fixed outputting error messages when recv returns with no data.
Also updated the comments to say that we're using non-blocking IO

Commit:
Added an example of how to send drive commands to a Husky from Windows

Commit:
Added copying the examples to CMakeLists for rosserial_windows

Commit:
Updated 64-bit float to be double on Windows.

Commit:
Updated sample code to be robot agnostic and fixed unitialized message
fields.

Commit:
Adding instructions for TestDrive example in rosserial_windows

kshehata committed:
Fixed numbered list and added link to tutorial

Commit:
Added an example of how to receive messages using rosserial_windows

Commit:
Merge branch 'winserial' of https://github.com/kshehata/rosserial into winserial

mikepurvis commented on rosserial_windows/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+catkin_package(CATKIN_DEPENDS)
+
+install(DIRECTORY src/ros_lib
+  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/src
```
This seems a bit hinky. Does it actually work from the installed version? ie, 

```
catkin_make install
source install/setup.bash
rosrun ...
```

mikeferguson replied:
This isn't source you would run with rosrun, this exactly line is used in all the rosserial libraries in order to copy over the source that is needed for the arduinos -- the make_library script then accesses that source.

mikepurvis commented on rosserial_windows/src/examples/MessageReceive/messagereceive.cpp:
```diff
@@ -0,0 +1,35 @@
+#include <stdio.h>
+#include "ros.h"
+#include <geometry_msgs/PoseWithCovarianceStamped.h>
+#include <windows.h>
```
What's the scheme here for quotes vs. angle-brackets? The style guide is silent on this type of thing, but please at least be consistent.

kshehata replied:
I've gone with quotes for anything that's in the project itself, and angle brackets for anything that is in include directories. Though ros.h would at first appear to be a system library is actually in the project, hence quotes.

mikepurvis commented on rosserial_windows/src/examples/MessageReceive/messagereceive.cpp:
```diff
@@ -0,0 +1,35 @@
+
+int _tmain(int argc, _TCHAR* argv[])
+{
+        ros::NodeHandle nh;
```
2 spaces instead of tabs, please.

kshehata replied:
Visual studio defaults to tabs. I was going for what would be most commonly used in Windows. I can change it to spaces, but anyone who goes to use in VS will likely fight with their editor.

mikepurvis replied:
Does it? Not the impression I get from this article: http://lostechies.com/jimmybogard/2012/03/07/tabs-versus-spaces-spaces-won/

It's good to cooperate with the local convention, but ROS style is still 2 spaces. A user building on rosserial in Windows is still creating ROS ecosystem code, so it really should be according to ROS conventions. Especially because a rosserial_windows user might not even necessarily be using MSVC— they could be on CodeBlocks, Eclipse, etc.

Whichever way it goes, a note on the wiki about it is warranted.

kshehata replied:
Good point. I've changed the style to match ROS guidelines using indent. There are plenty of examples on the web for how to change the defaults in VS to be not-stupid.

mikepurvis commented on rosserial_windows/src/examples/MessageReceive/messagereceive.cpp:
```diff
@@ -0,0 +1,35 @@
+#include "stdafx.h"
```
I know it's just an example, but there should still be a BSD title block on this file.

mikepurvis commented:
Made a couple of notes for stylistic fixes, but overall this is great— thanks, @kshehata.

Any concerns about this, @mikeferguson, @PaulBouchier ?

mikeferguson commented on rosserial_windows/package.xml:
```diff
@@ -0,0 +1,27 @@
+<package>
+  <name>rosserial_windows</name>
+  <version>0.0.0</version>
```
@mikepurvis this may need to be made 0.5.5 for the next release to work correctly?

mikepurvis replied:
Yes it will. Might make sense to change it post merge as part of the release preparation, not sure.

mikeferguson commented on rosserial_windows/src/ros_lib/WindowsSocket.cpp:
```diff
@@ -0,0 +1,233 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ * Copyright (c) 2014, Clearpath Robotics Inc.
```
Is Clearpath right here?

kshehata replied:
I basically rewrote the WindowsSocket code. Hence the Clearpath on this file.

mikepurvis replied:
Not sure—is there a standard about this? At present, rosserial_server is copyright Clearpath. A lot of stuff is copyright Willow, OSRF, individual contributors. It gets a bit murky with multiple packages in a single repo. @tfoote, opinions on how this type of thing should be handled?

mikeferguson commented on rosserial_windows/src/ros_lib/WindowsSocket.cpp:
```diff
@@ -0,0 +1,233 @@
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials provided
+ *    with the distribution.
+ *  * Neither the name of Willow Garage, Inc. nor the names of its
```
Willow here? yeah, check the attribution on all these license headers.

kshehata replied:
Gah. I copied and pasted license text, though I did re-write this file. I'll correct the license header.

mikepurvis replied:
Haha, whoops. @kshehata, we have a standard Clearpath BSD title block [here](https://github.com/clearpathrobotics/headers/blob/master/bsd/cpp).

mikeferguson commented:
Other than licensing and style, LGTM. Gotta say, this was never a use case I imagined when creating rosserial... cool idea.

Commit:
Fixed license file for WindowsSocket and ros to remove refs to Willow.
Changed make_libraries.py back to Willow Garage, since there are only
minor changes from the original.
Added license headers to the examples.

Commit:
Fixed indent style to match ROS guidelines

kshehata commented:
Yea, it's a pretty cool use of the tool @mikeferguson. Have to admin: I didn't actually come up with this myself. Some of the other guys here at Clearpath came up with it, and I just cleaned it up from a hack to distributable code.

Commit:
Changed licenses to Clearpath default

kshehata commented:
Just pushed the changes to the licenses and the style for the examples.

mikeferguson commented:
Also, it might make sense to squash these commits into one -- I presume intermediate versions do not work.

kshehata commented:
Most of the intermediate versions should work, though if it's easier on your end, I can certainly squash it into one big commit.

kshehata commented:
Just following up on this: do you want this redone as one big commit, or is the current format ok?
